storage/remote/wal_watcher.go

```diff
@@ -0,0 +1,565 @@
+				continue
+			}
+
+			if d < w.currentSegment {
+				w.lastCheckpoint = dir
+				// This potentially takes a long time, should we run it in another go routine?
+				err = w.readCheckpoint(w.lastCheckpoint)
+				if err != nil {
+					level.Error(w.logger).Log("err", err)
+				}
+			} else {
+				level.Info(w.logger).Log("msg", "current segment is behind the checkpoint, skipping reading of checkpoint for now", "current", fmt.Sprintf("%08d", w.currentSegment), "checkpoint", dir)
+			}
+
+		case <-segmentTicker.C:
+			// check if new segments exist
+			_, last, err := wl.Segments()
+			if err != nil {
+				level.Error(w.logger).Log("err", err)
+				continue
+			}
+
+			if last > w.currentSegment {
+				w.readSegment(reader)
+				level.Info(w.logger).Log("msg", "a new segment exists, we should start reading it", "current", fmt.Sprintf("%08d", w.currentSegment), "new", fmt.Sprintf("%08d", last))
+				return nil
+			}
+
+		case <-readTicker.C:
+			if err := w.readSegment(reader); err != nil {
+				level.Error(w.logger).Log("err", err)
+				return err
+			}
+		}
+	}
+}
+
+func (w *WALWatcher) readSegment(r *wal.LiveReader) error {
+	for r.Next() && !isClosed(w.quit) {
+		err := w.decodeRecord(r.Record())
+
+		// Intentionally skip over record decode errors.
+		if err != nil {
+			level.Error(w.logger).Log("err", err)
+		}
+	}
+	return r.Err()
+}
+
+func (w *WALWatcher) decodeRecord(rec []byte) error {
+	var (
+		dec     tsdb.RecordDecoder
+		series  []tsdb.RefSeries
+		samples []tsdb.RefSample
+	)
+	switch dec.Type(rec) {
+	case tsdb.RecordSeries:
+		series, err := dec.Series(rec, series[:0])
+		if err != nil {
+			w.recordDecodeFailsMetric.Inc()
+			return err
+		}
+		w.seriesReadMetric.Add(float64(len(series)))
+		w.writer.StoreSeries(series, w.currentSegment)
+
+	case tsdb.RecordSamples:
+		samples, err := dec.Samples(rec, samples[:0])
+		if err != nil {
+			w.recordDecodeFailsMetric.Inc()
+			return err
+		}
+		w.samplesReadMetric.Add(float64(len(samples)))
+		// Blocks  until the sample is sent to all remote write endpoints or closed (because enqueue blocks).
+		w.writer.Append(samples)
+
+	case tsdb.RecordTombstones:
+		w.tombstonesReadMetric.Add(float64(len(samples)))
+
+	case tsdb.RecordInvalid:
+		w.invalidReadMetric.Add(float64(len(samples)))
+		return errors.New("invalid record")
+
+	default:
+		w.recordDecodeFailsMetric.Inc()
+		return errors.New("unknown TSDB record type")
+	}
+	return nil
+}
+
+// Read all the series records from a Checkpoint directory.
+func (w *WALWatcher) readCheckpoint(checkpointDir string) error {
+	sr, err := wal.NewSegmentsReader(checkpointDir)
+	if err != nil {
+		return errors.Wrap(err, "open checkpoint")
+	}
+	defer sr.Close()
+
+	split := strings.Split(checkpointDir, ".")
+	if len(split) != 2 {
+		return errors.Errorf("checkpoint dir name is not in the right format: %s", checkpointDir)
+	}
+
+	i, err := strconv.Atoi(split[1])
+	if err != nil {
+		i = w.currentSegment - 1
+	}
```

Comment:
This will need some exposition as a comment, I don't understand it right now.

Reply:
I don't know what I'm doing here. We should probably be returning that error and ensuring `w.lastCheckpoint` wasn't updated.